### ISSUE TITLE
Occasional indefinite hangs when viewing a commit through magit with which-function-mode active

### ISSUE THREAD

**Issue by billsacks** [magit/magit#4602]
First, thank you very much for all of the work you do to develop and maintain magit! I have been enjoying it immensely, and I really appreciate all of the time and energy that you have put in to creating such an amazing git interface.

I experienced this bug using the latest Emacs 28 pretest build from https://emacsformacosx.com/builds. This is using a clean startup (via moving aside my .emacs.d for this test). For now, I am working around this issue by giving an explicit list of modes in `which-func-modes`, rather than using the default value of `t` for that variable. So this is not critical to my personal use and enjoyment of emacs and magit.

(1) Outside of emacs, clone the git repository where I have noticed this issue:
```
mkdir ~/temporary
git clone https://github.com/ESCOMP/CTSM.git ~/temporary/ctsm_temp
```

(2) Launch emacs

(3) Install magit:
M-x list-packages
Then install magit; I installed the latest current release, 3.3.0. (I have also seen this with the latest development magit version installed through melpa.)

(4) Turn on which-function-mode:
M-x which-function-mode

(5) Open a dired buffer for the CTSM git repository from step (1):
C-x C-f ~/temporary/ctsm_temp

(6) Open magit status buffer:
C-x g

(7) Try to view a particular commit:
M-x magit-show-commit [RET] 90741806a

Emacs now hangs indefinitely with CPU usage near 100%. C-g doesn't help.

Sending emacs a SIGUSR2 generates the following backtrace:

```
  magit-section-goto(#<magit-section nil [commit-message commitbuf] 302-[312-]356>)
  magit-section-backward()
  magit-section--backward-protected()
  magit-section--backward-find(#f(compiled-function () #<bytecode 0x11a398a59e8531fa>))
  magit-imenu--diff-prev-index-position-function()
  imenu-default-create-index-function()
  imenu--make-index-alist(t)
  which-function()
  which-func-update-1(#<window 3 on magit-revision: ctsm_temp>)
  which-func-update()
  apply(which-func-update nil)
  timer-event-handler([t 0 0 500000 t which-func-update nil idle 0 nil])
```

This problem doesn't show up for all commits, but I have seen it for multiple commits. I can't tell what is different about these problematic commits, but at least some of them (including the one given above) are older commits that were originally made in svn and then migrated to git.

If you skip step (4) (i.e., do not turn on which-function-mode), then the commit shows up quickly and emacs does not hang. If you then turn on which-function-mode at that point, emacs hangs at some point while scrolling through the commit (around line 200 or 220 of the buffer, though it has varied a bit the two times I tried it; there is nothing obviously different or strange to me about that point in the buffer).

I also tried this with emacs27.2. The problem does not appear there, but that seems to be because which-function-mode isn't active in the magit-revision buffer, even if it is active globally. When I disabled then re-enabled which-function-mode while in the magit-revision buffer, which-function-mode became active in that buffer and I got the same hang behavior. So it seems that the underlying issue exists in emacs27, but it has become visible and problematic in emacs28 because now which-function-mode is trying to work in magit buffers.

The following is the information generated from report-emacs-bug in case it's useful; my git version is `git version 2.32.0 (Apple Git-132)`

```
In GNU Emacs 28.0.91 (build 1, x86_64-apple-darwin18.7.0, NS appkit-1671.60 Version 10.14.6 (Build 18G95))
of 2022-01-19 built on builder10-14.lan
Windowing system distributor 'Apple', version 10.3.2113
System Description: macOS 12.2

Configured using:
'configure --with-ns '--enable-locallisppath=/Library/Application
Support/Emacs/${version}/site-lisp:/Library/Application
Support/Emacs/site-lisp' --with-modules'

Configured features:
ACL GMP GNUTLS JSON LIBXML2 MODULES NOTIFY KQUEUE NS PDUMPER THREADS
TOOLKIT_SCROLL_BARS ZLIB

Important settings:
value of $LANG: en_US.UTF-8
locale-coding-system: utf-8-unix

Major mode: Magit Rev

Minor modes in effect:
goto-address-mode: t
bug-reference-mode: t
global-git-commit-mode: t
magit-auto-revert-mode: t
shell-dirtrack-mode: t
tooltip-mode: t
global-eldoc-mode: t
show-paren-mode: t
electric-indent-mode: t
mouse-wheel-mode: t
tool-bar-mode: t
menu-bar-mode: t
file-name-shadow-mode: t
global-font-lock-mode: t
font-lock-mode: t
blink-cursor-mode: t
auto-composition-mode: t
auto-encryption-mode: t
auto-compression-mode: t
buffer-read-only: t
line-number-mode: t
indent-tabs-mode: t
transient-mark-mode: t

Load-path shadows:
/Users/sacks/.emacs.d/elpa/transient-0.3.7/transient hides /Applications/Emacs-alt.app/Contents/Resources/lisp/transient

Features:
(shadow sort mail-extr emacsbug sendmail magit-extras goto-addr vc-mtn
vc-hg vc-git vc-bzr vc-src vc-sccs vc-svn vc-cvs vc-rcs vc vc-dispatcher
bug-reference thingatpt face-remap magit-submodule magit-obsolete
magit-blame magit-stash magit-reflog magit-bisect magit-push magit-pull
magit-fetch magit-clone magit-remote magit-commit magit-sequence
magit-notes magit-worktree magit-tag magit-merge magit-branch
magit-reset magit-files magit-refs magit-status magit magit-repos
magit-apply magit-wip magit-log which-func imenu magit-diff smerge-mode
diff diff-mode git-commit rx log-edit easy-mmode message rmc puny rfc822
mml mml-sec epa derived epg rfc6068 epg-config gnus-util rmail
rmail-loaddefs text-property-search time-date mm-decode mm-bodies
mm-encode mail-parse rfc2231 rfc2047 rfc2045 mm-util ietf-drums
mail-prsvr mailabbrev mail-utils gmm-utils mailheader pcvs-util add-log
magit-core magit-autorevert autorevert filenotify magit-margin
magit-transient magit-process with-editor shell pcomplete comint ring
server ansi-color magit-mode transient cl-extra help-mode format-spec
magit-git magit-section magit-utils crm dash dired-aux dired
dired-loaddefs edmacro kmacro info package browse-url url url-proxy
url-privacy url-expand url-methods url-history url-cookie url-domsuf
url-util mailcap url-handlers url-parse auth-source cl-seq eieio
eieio-core cl-macs eieio-loaddefs password-cache json subr-x map
url-vars seq byte-opt gv bytecomp byte-compile cconv cl-loaddefs cl-lib
iso-transl tooltip eldoc paren electric uniquify ediff-hook vc-hooks
lisp-float-type elisp-mode mwheel term/ns-win ns-win ucs-normalize
mule-util term/common-win tool-bar dnd fontset image regexp-opt fringe
tabulated-list replace newcomment text-mode lisp-mode prog-mode register
page tab-bar menu-bar rfn-eshadow isearch easymenu timer select
scroll-bar mouse jit-lock font-lock syntax font-core term/tty-colors
frame minibuffer cl-generic cham georgian utf-8-lang misc-lang
vietnamese tibetan thai tai-viet lao korean japanese eucjp-ms cp51932
hebrew greek romanian slovak czech european ethiopic indian cyrillic
chinese composite emoji-zwj charscript charprop case-table epa-hook
jka-cmpr-hook help simple abbrev obarray cl-preloaded nadvice button
loaddefs faces cus-face macroexp files window text-properties overlay
sha1 md5 base64 format env code-pages mule custom widget
hashtable-print-readable backquote threads kqueue cocoa ns multi-tty
make-network-process emacs)

Memory information:
((conses 16 137612 11147)
(symbols 48 14293 1)
(strings 32 49535 1429)
(string-bytes 1 1593959)
(vectors 16 31096)
(vector-slots 8 367333 21454)
(floats 8 69 98)
(intervals 56 1207 311)
(buffers 992 14))
```

**Comment by kyleam** [comment]
Thanks for the detailed report.

> This is using a clean startup (via moving aside my .emacs.d for this test).

Fwiw, you can keep your `~/.emacs.d/` in place and start emacs with `-q` or `-Q` (depending on what you want) for this sort of testing.  Then you can point `--load` to a custom file that loads the minimal set of things you need for the test.

Also, Magit's Makefile also has an `emacs-Q` target that you may find helpful here (though I can't say much more about that becaus I don't use it myself).

> (1) Outside of emacs, clone the git repository where I have noticed this issue:
> 
> ```
> mkdir ~/temporary
> git clone https://github.com/ESCOMP/CTSM.git ~/temporary/ctsm_temp
> ```
> [...]
> 
> (4) Turn on which-function-mode: M-x which-function-mode
> [...]
> (7) Try to view a particular commit: M-x magit-show-commit [RET] 90741806a
> 
> Emacs now hangs indefinitely with CPU usage near 100%. C-g doesn't help.

I tried on a GNU/Linux system with Magit v3.3.0-149-gf0296148 and Emacs 28.0.91-87-gb7a651ba378 and was able to trigger a `which-function-mode`-dependent CPU spike and what looks like a hang.  (`C-g` can interrupt it in my case.)

> I also tried this with emacs27.2. The problem does not appear there, but that seems to be because which-function-mode isn't active in the magit-revision buffer, even if it is active globally. [...]

Indeed Emacs 27.2 seems okay on my end too.  Looking through the new commits to which-func.el, bd64571ef21 (which-function-mode: put hook function on after-change-major-mode-hook, 2020-05-20) seems like a good candidate for the change in behavior.  After reverting that, I don't seem the problem.  That commit was in response to https://debbugs.gnu.org/40714 ("Reinitialising the major mode disables which-function-mode"), which sounds like it fits with what you're saying about `which-function-mode` not actually being active for Emacs 27.

I haven't debugged any further, but I agree that, given the current details, the Magit issue tracker seems like the right place for this.

**Comment by kyleam** [comment]
I believe the core issue is having something that `magit-insert-revision-message` will convert to a `(commit text)` section at the start of the `(commit-message)` section.  If this is true, you should no longer be able to trigger the problem when `magit-revision-use-hash-sections` is nil.

Here's a minimal reproducer:

```sh
set -eu

tdir="$(mktemp -d "${TMPDIR:-/tmp}"/magit-4602-XXXXXXX)"
cd "$tdir"
git init
git commit -mc0 --allow-empty
git commit -m"$(git rev-parse HEAD)" --allow-empty
# then magit-show-commit HEAD
```

Will work on a fix in a bit.

**Comment by kyleam** [comment]
Should be fixed by e47973968.  @billsacks, please re-open if that doesn't resolve these hangs on your end.

**Comment by billsacks** [comment]
Thank you very much!
